File: backend/modules/Reminders/Wathiq.Reminders.Domain/Reminders/Reminder.cs
using System;
using System.Collections.Generic;
using Volo.Abp.Domain.Entities.Auditing;
using Wathiq.Reminders.Rules;

namespace Wathiq.Reminders.Reminders;

/// <summary>
/// One scheduled reminder: document × offset (FR-REM-001). The pair is database-unique
/// (UQ_Reminder_DocumentId_OffsetDays), so rescheduling REUSES this row via Reschedule()
/// instead of delete-and-insert - that reuse is the backbone of FR-REM-002 idempotency.
/// </summary>
public class Reminder : FullAuditedAggregateRoot<Guid>
{
    public Guid UserId { get; private set; }
    public Guid DocumentId { get; private set; }   // -> documents.Document by id only, no FK (DB2)
    public int OffsetDays { get; private set; }
    public DateOnly DueDate { get; private set; }
    public ReminderStatus Status { get; private set; }
    public DateTime? SentAt { get; private set; }  // UTC

    private readonly List<DeliveryLog> _deliveryLogs = new();
    public IReadOnlyCollection<DeliveryLog> DeliveryLogs => _deliveryLogs.AsReadOnly();

    private Reminder()
    {
    }

    public Reminder(Guid id, Guid userId, Guid documentId, int offsetDays, DateOnly dueDate)
        : base(id)
    {
        UserId = userId;
        DocumentId = documentId;
        OffsetDays = offsetDays;
        DueDate = dueDate;
        Status = ReminderStatus.Pending;
    }

    /// <summary>Re-arms the row for a new due date (expiry changed / document renewed). SentAt clears; history stays in DeliveryLogs.</summary>
    public Reminder Reschedule(DateOnly newDueDate)
    {
        DueDate = newDueDate;
        Status = ReminderStatus.Pending;
        SentAt = null;
        return this;
    }

    /// <summary>A sent reminder stays Sent - cancelling only stops ones that haven't gone out.</summary>
    public Reminder Cancel()
    {
        if (Status is ReminderStatus.Pending or ReminderStatus.Failed)
        {
            Status = ReminderStatus.Cancelled;
        }

        return this;
    }

    /// <summary>The only way a DeliveryLog is born; also the Pending -> Sent/Failed transition (FR-REM-002/005).
    /// The id comes from the caller (GuidGenerator - sequential, DB3), same as Document.AddAttachment.</summary>
    public DeliveryLog RecordAttempt(Guid logId, ReminderChannels channel, DateTime attemptedAtUtc, bool succeeded, string? error = null)
    {
        var log = new DeliveryLog(logId, Id, channel, attemptedAtUtc, succeeded, error);
        _deliveryLogs.Add(log);

        if (succeeded)
        {
            Status = ReminderStatus.Sent;
            SentAt = attemptedAtUtc;
        }
        else
        {
            Status = ReminderStatus.Failed;
        }

        return log;
    }
}

File: mobile/lib/core/models/reminder.dart
/// Reminder models (consumed from 6.9): system-managed rows, read-only here.
library;

enum ReminderStatus {
  pending(0),
  sent(1),
  failed(2),
  cancelled(3);

  const ReminderStatus(this.wire);
  final int wire;

  static ReminderStatus fromWire(int value) =>
      values.firstWhere((s) => s.wire == value);
}

class Reminder {
  const Reminder({
    required this.id,
    required this.documentId,
    required this.offsetDays,
    required this.dueDate,
    required this.expiryDate,
    required this.status,
    this.sentAt,
  });

  final String id;
  final String documentId;
  final int offsetDays;
  final DateTime dueDate;
  final DateTime expiryDate;
  final ReminderStatus status;
  final DateTime? sentAt;   // null until dispatched - history is immutable (2.x)

  factory Reminder.fromJson(Map<String, dynamic> json) => Reminder(
        id: json['id'] as String,
        documentId: json['documentId'] as String,
        offsetDays: json['offsetDays'] as int,
        dueDate: DateTime.parse(json['dueDate'] as String),
        expiryDate: DateTime.parse(json['expiryDate'] as String),
        status: ReminderStatus.fromWire(json['status'] as int),
        sentAt: json['sentAt'] == null
            ? null
            : DateTime.parse(json['sentAt'] as String),
      );
}
